refactor: removed previously deprecated -bpf-lb-proto-diff option.

Surya-7890 opened the pull request:
This commit removes -bpf-lb-proto-diff, that was marked as deprecated in v1.18.

Please ensure your pull request adheres to the following guidelines:
- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #39328

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -337,6 +337,7 @@ Removed Options
 * The previously deprecated ``enableRuntimeDeviceDetection`` option has been removed
 * The previously deprecated and ignored operator flags ``ces-write-qps-limit``, ``ces-write-qps-burst``, ``ces-enable-dynamic-rate-limit``,
   ``ces-dynamic-rate-limit-nodes``, ``ces-dynamic-rate-limit-qps-limit``, ``ces-dynamic-rate-limit-qps-burst`` have been removed.
+* The previously deprecated ``--bpf-lb-proto-diff`` flag has been removed.
```
This would be the first upgrade note of the v1.19 cycle. Hence, all previous notes should be removed (in a separate commit), and the version should be updated to 1.19.

Surya-7890 replied:
Removed the notes for `v1.18` from version specific notes, leaving only this note. Please take a look to make sure, if  this is the expected outcome.

giorio94 replied:
Yes, makes sense to me. Could you please have this one as the first commit to remove all old entries (updating the commit message accordingly), and then move the addition of the *"The previously deprecated ``--bpf-lb-proto-diff`` flag has been removed"* entry to the other commit alongside the other changes.

Surya-7890 replied:
I am thinking, I should change the version to 1.19 in the initial commit. Would like to know get your opinion on this, please.

giorio94 replied:
Do you mean in the upgrade notes? Yeah, makes sense to me. Basically the idea is just to have a preparatory commit for the upgrade notes, and then one that actually does implement the removal of the option (and updates the upgrade notes with that info).

Surya-7890 replied:
No, I meant changing the version of upgrade notes to v1.19 and removing the notes for v1.18 in a single commit and another commit that actually implements the removal of the option.

giorio94 replied:
Yep, that is the structure I was mentioning as well.

Surya-7890 replied:
I updated the commit structure. PTAL when you get a chance!

giorio94 replied:
Looks perfect, thanks

giorio94 commented on pkg/datapath/linux/config/config.go:
```diff
@@ -707,10 +707,6 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	// Write Identity and ClusterID related macros.
 	cDefinesMap["CLUSTER_ID_MAX"] = fmt.Sprintf("%d", option.Config.MaxConnectedClusters)
 
-	if option.Config.LoadBalancerProtocolDifferentiation {
-		cDefinesMap["ENABLE_SERVICE_PROTOCOL_DIFFERENTIATION"] = "1"
-	}
```
All references to `ENABLE_SERVICE_PROTOCOL_DIFFERENTIATION` inside the datapath should be removed as well.

Surya-7890 replied:
All other instances of `ENABLE_SERVICE_PROTOCOL_DIFFERENTIATION` were inside bpf. I don't know if they are related somehow, should i remove all those references as well?

giorio94 replied:
Yes they are related. That macro used to be set to `1` here, so given that you removed this assignment you also need to drop all the other occurrences, preserving the behavior as if it were set.

Surya-7890 replied:
I added a commit that removes all other occurrences of `ENABLE_SERVICE_PROTOCOL_DIFFERENTIATION` and another one for the the removal of version specific notes.

giorio94 commented on pkg/k8s/watchers/service.go:
```diff
@@ -513,12 +513,6 @@ func (k *K8sServiceWatcher) addK8sSVCs(svcID k8s.ServiceID, oldSvc, svc *k8s.Ser
-	if !option.Config.LoadBalancerProtocolDifferentiation {
-		oldSvc = stripServiceProtocol(oldSvc)
-		svc = stripServiceProtocol(svc)
-		endpoints = stripEndpointsProtocol(endpoints)
```
The `stripEndpointsProtocol` looks now unused, and should be removed.

Surya-7890 replied:
When i rebased, I noticed that `pkg/k8s/watchers/service.go` and `pkg/k8s/watchers/service_test.go` files were deleted in this [PR](https://github.com/cilium/cilium/commit/fa0e279d1884d0f522d7859c7d1d1c6e24bcbf1b).

And the function `stripEndpointsProtocol`, which was in `pkg/k8s/watchers/service.go` , was also deleted. I am force pushing the rebased code.

giorio94 reviewed (changes requested):
Thanks @Surya-7890 for taking care of this. I've left a few initial comments inline about additional clean-up that should be performed. Moreover, you should also rebase this PR onto the main branch to address the conflicts.

tommyp1ckles commented on pkg/option/config.go:
```diff
@@ -3060,7 +3049,6 @@ func (c *DaemonConfig) Populate(logger *slog.Logger, vp *viper.Viper) {
 		c.IdentityRestoreGracePeriod = defaults.IdentityRestoreGracePeriodKvstore
 	}
 
-	c.LoadBalancerProtocolDifferentiation = vp.GetBool(LoadBalancerProtocolDifferentiation)
```
You'll probably have to regenerate the documentation following this.

giorio94 replied:
This (the regeneration of the documentation) is probably not required (CI doesn't seem to complain) because the flag was already marked as deprecated.

Surya-7890 replied:
If this has to be updated, I'll try to do this in this pr. Just to confirm, is the command to generate documentation, `make docs` inside `install/kubernetes`?

giorio94 replied:
The command is `make -C Documentation update-cmdref`, but I wouldn't expect any changes, as CI would have otherwise complained.

tommyp1ckles commented on pkg/option/config.go:
```diff
@@ -3060,7 +3049,6 @@ func (c *DaemonConfig) Populate(logger *slog.Logger, vp *viper.Viper) {
 		c.IdentityRestoreGracePeriod = defaults.IdentityRestoreGracePeriodKvstore
 	}
 
-	c.LoadBalancerProtocolDifferentiation = vp.GetBool(LoadBalancerProtocolDifferentiation)
```
Also, the flag needs to be removed from the helm config map: [install/kubernetes/cilium/templates/cilium-configmap.yaml ](https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-configmap.yaml#L839-L841)

Surya-7890 replied:
I removed it in my recent commit. PTAL when possible.

giorio94 reviewed (commented):
Thanks! Changes look good to me. Just one cosmetic comment about the commits structure, in addition to Tom's. I'll trigger the full test suite once these are addressed.

Surya-7890 committed:
changelog: remove v1.18-specific notes, update heading to v1.19

This commit removes notes that are specific to v1.18.
The section heading has also been updated to v1.19.

Signed-off-by: surya <suryamariappan9108@gmail.com>

Surya-7890 committed:
refactor: removed previously deprecated -bpf-lb-proto-diff option.

Fixes: #39328

This commit removes the -bpf-lb-proto-diff option, which was marked as deprecated in v1.18.

Signed-off-by: surya <suryamariappan9108@gmail.com>

giorio94 reviewed (approved):
/lgtm, thanks!

tommyp1ckles reviewed (approved):
k8s changes lgtm, thanks !

aditighag reviewed (approved):
Thanks!
